syntax = "proto3";
package slog;

enum TransactionType {
    UNKNOWN = 0;
    SINGLE_HOME = 1;
    MULTI_HOME = 2;
}

message MasterMetadata {
    uint32 master = 1;
    uint32 counter = 2;
}

message TransactionInternal {
    uint32 id = 1;
    map<string, MasterMetadata> master_metadata = 2;
    TransactionType type = 3;
}

message Transaction {
    map<string, string> read_set = 1;
    map<string, string> write_set = 2;
    string code = 3;

    TransactionInternal internal = 4;
}